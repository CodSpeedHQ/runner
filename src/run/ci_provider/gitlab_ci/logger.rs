use console::style;
use lazy_static::lazy_static;
use log::{Level, LevelFilter, Log};
use regex::Regex;
use simplelog::SharedLogger;
use std::{
    env,
    io::Write,
    sync::Mutex,
    time::{SystemTime, UNIX_EPOCH},
};

use crate::{
    logger::{get_group_event, GroupEvent},
    run::ci_provider::logger::should_provider_logger_handle_record,
};

lazy_static! {
    static ref GITLAB_SECTION_ID_SANITIZE_REGEX: Regex =
        Regex::new(r"[^\d\w\-_]").expect("Failed to compile GitLab SectionId regex");
}

/// Unicode Escape character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_ESC: char = '\x1B';

/// Unicode Carriage Return character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_CR: char = '\x0D';

/// Reset color mode
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
const COLOR_RESET: &str = "\x1B[0m";

/// Erase cursor till end of line
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#erase-functions
const ERASE_CURSOR: &str = "\x1B[0K";

/// A logger that prints log in the format expected by GitLab CI
///
/// See https://docs.gitlab.com/ee/ci/yaml/script.html
pub struct GitLabCILogger {
    log_level: LevelFilter,
    section_id: Mutex<Option<String>>,
}

impl GitLabCILogger {
    pub fn new() -> Self {
        // force activation of colors, because GitlabCI does not
        // respect the CLICOLORS spec.
        // https://gitlab.com/gitlab-org/gitlab/-/issues/28598
        console::set_colors_enabled(true);

        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self {
            log_level,
            section_id: Mutex::new(None),
        }
    }
}

impl Log for GitLabCILogger {
    fn enabled(&self, _metadata: &log::Metadata) -> bool {
        true
    }

    fn log(&self, record: &log::Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            let now = SystemTime::now();
            let timestamp = now.duration_since(UNIX_EPOCH).unwrap().as_secs();
            let mut section_id = self.section_id.lock().unwrap();

            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    let new_section_id = GITLAB_SECTION_ID_SANITIZE_REGEX
                        .replace_all(&name, "_")
                        .to_ascii_lowercase();

                    *section_id = Some(new_section_id.to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_start:{timestamp}:{new_section_id}{U_CR}{ERASE_CURSOR}{U_ESC}[36;1m{name}{COLOR_RESET}");
                }
                GroupEvent::End => {
                    // do not fail if there is no current section
                    let current_section_id = section_id.clone().unwrap_or("".to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_end:{timestamp}:{current_section_id}{U_CR}{ERASE_CURSOR}");

                    *section_id = None;
                }
            }
            return;
        }

        if level > self.log_level {
            return;
        }

        // set log colors. See https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
        match level {
            Level::Error => {
                println!("{}", style(message).red());
            }
            Level::Warn => {
                println!("{}", style(message).yellow());
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("{}", style(message).cyan());
            }
            Level::Trace => {
                println!("{}", style(message).magenta());
            }
        };
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GitLabCILogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
