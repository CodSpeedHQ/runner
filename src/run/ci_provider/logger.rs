use crate::run::runner::VALGRIND_EXECUTION_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the valgrind execution target
    record.target() != VALGRIND_EXECUTION_TARGET
}
