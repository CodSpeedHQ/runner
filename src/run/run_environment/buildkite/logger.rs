use crate::{
    logger::{get_group_event, get_json_event, GroupEvent},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by Buildkite
///
/// See https://buildkite.com/docs/pipelines/managing-log-output
pub struct BuildkiteLogger {
    log_level: LevelFilter,
}

impl BuildkiteLogger {
    pub fn new() -> Self {
        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self { log_level }
    }
}

impl Log for BuildkiteLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) => {
                    println!("--- {name}");
                }
                GroupEvent::StartOpened(name) => {
                    println!("+++ {name}");
                }
                GroupEvent::End => {}
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }
        // there is no support for log levels in Buildkite, so we print the level in the message
        match level {
            Level::Error => {
                println!("[ERROR] {message}");
            }
            Level::Warn => {
                println!("[WARN] {message}");
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("[DEBUG] {message}");
            }
            Level::Trace => {
                println!("[TRACE] {message}");
            }
        }
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for BuildkiteLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
