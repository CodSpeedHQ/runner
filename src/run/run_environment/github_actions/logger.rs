use crate::{
    logger::{GroupEvent, get_group_event, get_json_event},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by GitHub Actions, with grouping support.
///
/// See https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions
pub struct GithubActionLogger {
    log_level: LevelFilter,
}

impl GithubActionLogger {
    pub fn new() -> Self {
        // Only enable debug logging if it's enabled in GitHub Actions.
        // See: https://docs.github.com/en/actions/reference/workflows-and-actions/variables
        let log_level = if env::var("RUNNER_DEBUG").unwrap_or_default() == "1" {
            LevelFilter::Trace
        } else {
            env::var("CODSPEED_LOG")
                .ok()
                .and_then(|log_level| log_level.parse::<LevelFilter>().ok())
                .unwrap_or(LevelFilter::Info)
        };

        Self { log_level }
    }
}

impl Log for GithubActionLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    println!("::group::{name}");
                }
                GroupEvent::End => {
                    println!("::endgroup::");
                }
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }

        let prefix = match level {
            Level::Error => "::error::",
            Level::Warn => "::warning::",
            Level::Info => "",
            Level::Debug => "::debug::",
            Level::Trace => "::debug::[TRACE]",
        };
        let message_string = message.to_string();
        let lines = message_string.lines();
        // ensure that all the lines of the message have the prefix, otherwise GitHub Actions will not recognize the command for the whole string
        lines.for_each(|line| println!("{prefix}{line}"));
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GithubActionLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
