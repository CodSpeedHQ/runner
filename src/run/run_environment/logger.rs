use crate::run::runner::EXECUTOR_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the executor target
    record.target() != EXECUTOR_TARGET
}
