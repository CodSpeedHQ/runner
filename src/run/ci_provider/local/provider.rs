use git2::Repository;
use lazy_static::lazy_static;
use simplelog::SharedLogger;

use crate::logger::get_local_logger;
use crate::prelude::*;
use crate::run::{
    ci_provider::{
        interfaces::{ProviderMetadata, RunEvent},
        provider::{CIProvider, CIProviderDetector},
    },
    config::Config,
    helpers::find_repository_root,
};

#[derive(Debug)]
pub struct LocalProvider {
    pub ref_: String,
    pub owner: String,
    pub repository: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub event: RunEvent,
    pub repository_root_path: String,
}

impl LocalProvider {}

lazy_static! {
    static ref REMOTE_REGEX: regex::Regex =
        regex::Regex::new(r"[:/](?P<owner>[^/]+)/(?P<repository>[^/]+)\.git").unwrap();
}

fn extract_owner_and_repository_from_remote_url(remote_url: &str) -> Result<(String, String)> {
    let captures = REMOTE_REGEX.captures(remote_url).ok_or_else(|| {
        anyhow!(
            "Could not extract owner and repository from remote url: {}",
            remote_url
        )
    })?;

    let owner = captures.name("owner").unwrap().as_str();
    let repository = captures.name("repository").unwrap().as_str();

    Ok((owner.to_string(), repository.to_string()))
}

impl TryFrom<&Config> for LocalProvider {
    type Error = Error;
    fn try_from(_config: &Config) -> Result<Self> {
        let repository_root_path = match find_repository_root(&std::env::current_dir()?) {
            Some(mut path) => {
                // Add a trailing slash to the path
                path.push("");
                path.to_string_lossy().to_string()
            },
            None => bail!("Could not find repository root, please make sure you are running the command from inside a git repository"),
        };

        let git_repository = Repository::open(repository_root_path.clone()).context(format!(
            "Failed to open repository at path: {}",
            repository_root_path
        ))?;

        let remote = git_repository.find_remote("origin")?;
        let (owner, repository) =
            extract_owner_and_repository_from_remote_url(remote.url().unwrap())?;

        let head = git_repository.head().context("Failed to get HEAD")?;
        let ref_ = head
            .peel_to_commit()
            .context("Failed to get HEAD commit")?
            .id()
            .to_string();
        let head_ref = if head.is_branch() {
            let branch = head.shorthand().context("Failed to get HEAD branch name")?;
            Some(branch.to_string())
        } else {
            None
        };

        Ok(Self {
            ref_,
            head_ref,
            base_ref: None,
            owner,
            repository,
            event: RunEvent::Local,
            repository_root_path,
        })
    }
}

impl CIProviderDetector for LocalProvider {
    fn detect() -> bool {
        true
    }
}

impl CIProvider for LocalProvider {
    fn get_logger(&self) -> Box<dyn SharedLogger> {
        get_local_logger()
    }

    fn get_provider_name(&self) -> &'static str {
        "Local"
    }

    fn get_provider_slug(&self) -> &'static str {
        "local"
    }

    fn get_provider_metadata(&self) -> Result<ProviderMetadata> {
        Ok(ProviderMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            event: self.event.clone(),
            gh_data: None,
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),
        })
    }
}

#[cfg(test)]
mod tests {
    // use crate::VERSION;
    // use insta::assert_json_snapshot;

    use super::*;

    #[test]
    fn test_extract_owner_and_repository_from_remote_url() {
        let remote_urls = [
            "git@github.com:CodSpeedHQ/runner.git",
            "https://github.com/CodSpeedHQ/runner.git",
        ];
        for remote_url in remote_urls.iter() {
            let (owner, repository) =
                extract_owner_and_repository_from_remote_url(remote_url).unwrap();
            assert_eq!(owner, "CodSpeedHQ");
            assert_eq!(repository, "runner");
        }
    }

    // TODO: uncomment later when we have a way to mock git repository
    // #[test]
    // fn test_provider_metadata() {
    //     let config = Config {
    //         token: Some("token".into()),
    //         ..Config::test()
    //     };
    //     let local_provider = LocalProvider::try_from(&config).unwrap();
    //     let provider_metadata = local_provider.get_provider_metadata().unwrap();

    //     assert_json_snapshot!(provider_metadata, {
    //         ".runner.version" => insta::dynamic_redaction(|value,_path| {
    //             assert_eq!(value.as_str().unwrap(), VERSION.to_string());
    //             "[version]"
    //         }),
    //     });
    // }
}
