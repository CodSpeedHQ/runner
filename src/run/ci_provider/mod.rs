pub mod interfaces;
pub mod logger;
mod provider;

use buildkite::BuildkiteProvider;
use github_actions::GitHubActionsProvider;
use local::LocalProvider;
use provider::CIProviderDetector;

use crate::prelude::*;
use crate::run::config::Config;

pub use self::provider::CIProvider;

// Provider implementations
mod buildkite;
mod github_actions;
mod local;

pub fn get_provider(config: &Config) -> Result<Box<dyn CIProvider>> {
    if BuildkiteProvider::detect() {
        let provider = BuildkiteProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitHubActionsProvider::detect() {
        let provider = GitHubActionsProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if LocalProvider::detect() {
        let provider = LocalProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    bail!("No CI provider detected")
}
