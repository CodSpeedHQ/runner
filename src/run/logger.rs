use crate::logger::{GROUP_TARGET, OPENED_GROUP_TARGET};
use crate::prelude::*;
use crate::run::{ci_provider::CIProvider, runner::RunData};
use log::LevelFilter;
use simplelog::{CombinedLogger, WriteLogger};
use std::fs::copy;
use std::path::PathBuf;
use tempfile::NamedTempFile;

pub struct Logger {
    log_file_path: PathBuf,
}

impl Logger {
    #[allow(clippy::borrowed_box)]
    pub fn new(provider: &Box<dyn CIProvider>) -> Result<Self> {
        let provider_logger = provider.get_logger();
        let log_file = NamedTempFile::new().context("Failed to create log file")?;
        let log_file_path = log_file.path().to_path_buf();
        let file_logger_config = simplelog::ConfigBuilder::new()
            // Groups are not logged to the file
            .add_filter_ignore_str(GROUP_TARGET)
            .add_filter_ignore_str(OPENED_GROUP_TARGET)
            .build();
        let file_logger = WriteLogger::new(LevelFilter::Trace, file_logger_config, log_file);
        CombinedLogger::init(vec![provider_logger, file_logger])
            .context("Failed to init logger")?;
        Ok(Self { log_file_path })
    }

    pub fn persist_log_to_profile_folder(&self, run_data: &RunData) -> Result<()> {
        let profile_folder = run_data.profile_folder.clone();
        let dest_log_file_path = profile_folder.join("runner.log");
        debug!("Persisting log file to {}", dest_log_file_path.display());
        log::logger().flush();
        copy(&self.log_file_path, dest_log_file_path).context("Failed to copy log file")?;
        Ok(())
    }
}
