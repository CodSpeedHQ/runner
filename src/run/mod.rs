use crate::api_client::CodSpeedAPIClient;
use crate::config::CodSpeedConfig;
use crate::prelude::*;
use crate::run::{config::Config, logger::Logger};
use crate::VERSION;
use check_system::SystemInfo;
use clap::Args;
use instruments::mongo_tracer::MongoTracer;
use runner::get_run_data;

mod check_system;
pub mod ci_provider;
mod helpers;
mod instruments;
mod poll_results;
mod runner;
mod uploader;

pub mod config;
pub mod logger;

fn show_banner() {
    let banner = format!(
        r#"
   ______            __ _____                         __
  / ____/____   ____/ // ___/ ____   ___   ___   ____/ /
 / /    / __ \ / __  / \__ \ / __ \ / _ \ / _ \ / __  /
/ /___ / /_/ // /_/ / ___/ // /_/ //  __//  __// /_/ /
\____/ \____/ \__,_/ /____// .___/ \___/ \___/ \__,_/
  https://codspeed.io     /_/          runner v{}
"#,
        VERSION
    );
    println!("{}", banner);
    debug!("codspeed v{}", VERSION);
}

#[derive(Args, Debug)]
pub struct RunArgs {
    /// The upload URL to use for uploading the results, useful for on-premises installations
    #[arg(long)]
    pub upload_url: Option<String>,

    /// The token to use for uploading the results,
    #[arg(long, env = "CODSPEED_TOKEN")]
    pub token: Option<String>,

    /// The directory where the command will be executed.
    #[arg(long)]
    pub working_directory: Option<String>,

    /// Comma-separated list of instruments to enable. Possible values: mongodb.
    #[arg(long, value_delimiter = ',')]
    pub instruments: Vec<String>,

    /// The name of the environment variable that contains the MongoDB URI to patch.
    /// If not provided, user will have to provide it dynamically through a CodSpeed integration.
    ///
    /// Only used if the `mongodb` instrument is enabled.
    #[arg(long)]
    pub mongo_uri_env_name: Option<String>,

    /// Only for debugging purposes, skips the upload of the results
    #[arg(
        long,
        default_value = "false",
        hide = true,
        env = "CODSPEED_SKIP_UPLOAD"
    )]
    pub skip_upload: bool,

    /// Only for debugging purposes, skips the setup of the runner
    #[arg(long, default_value = "false", hide = true)]
    pub skip_setup: bool,

    /// The bench command to run
    pub command: Vec<String>,
}

#[cfg(test)]
impl RunArgs {
    /// Constructs a new `RunArgs` with default values for testing purposes
    pub fn test() -> Self {
        Self {
            upload_url: None,
            token: None,
            working_directory: None,
            instruments: vec![],
            mongo_uri_env_name: None,
            skip_upload: false,
            skip_setup: false,
            command: vec![],
        }
    }
}

pub async fn run(args: RunArgs, api_client: &CodSpeedAPIClient) -> Result<()> {
    let mut config = Config::try_from(args)?;
    let provider = ci_provider::get_provider(&config)?;
    let codspeed_config = CodSpeedConfig::load()?;
    let logger = Logger::new(&provider)?;

    if provider.get_provider_slug() != "local" {
        show_banner();
    }
    debug!("config: {:#?}", config);

    if provider.get_provider_slug() == "local" {
        if codspeed_config.auth.token.is_none() {
            bail!("You have to authenticate the CLI first. Run `codspeed auth login`.");
        }
        debug!("Using the token from the CodSpeed configuration file");
        config.set_token(codspeed_config.auth.token.clone());
    }

    let system_info = SystemInfo::new()?;
    check_system::check_system(&system_info)?;

    let mode = runner::get_mode()?;
    let executor = runner::get_executor_from_mode(mode);

    let run_data = get_run_data()?;

    if !config.skip_setup {
        start_group!("Preparing the environment");
        executor.setup(&config, &system_info, &run_data).await?;
        end_group!();
    }

    start_opened_group!("Running the benchmarks");

    // TODO: refactor and move directly in the Instruments struct as a `start` method
    let mongo_tracer = if let Some(mongodb_config) = &config.instruments.mongodb {
        let mut mongo_tracer = MongoTracer::try_from(&run_data.profile_folder, mongodb_config)?;
        mongo_tracer.start().await?;
        Some(mongo_tracer)
    } else {
        None
    };

    executor
        .run(&config, &system_info, &run_data, &mongo_tracer)
        .await?;

    // TODO: refactor and move directly in the Instruments struct as a `stop` method
    if let Some(mut mongo_tracer) = mongo_tracer {
        mongo_tracer.stop().await?;
    }

    executor.teardown(&config, &system_info, &run_data).await?;

    end_group!();

    if !config.skip_upload {
        start_group!("Uploading performance data");
        logger.persist_log_to_profile_folder(&run_data)?;
        let upload_result =
            uploader::upload(&config, &system_info, &provider, &run_data, executor.name()).await?;
        end_group!();

        if provider.get_provider_slug() == "local" {
            start_group!("Fetching the results");
            poll_results::poll_results(api_client, &provider, upload_result.run_id).await?;
            end_group!();
        }
    }

    Ok(())
}
