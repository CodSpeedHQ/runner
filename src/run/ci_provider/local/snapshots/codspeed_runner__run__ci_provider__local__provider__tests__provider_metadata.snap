---
source: src/run/ci_provider/local/provider.rs
expression: provider_metadata
---
{
  "ref": "18ec1d64b5f25fb27451d89eee03cc569bd6bbb1",
  "headRef": "feat/branch",
  "owner": "my-org",
  "repository": "adrien-python-test",
  "event": "local",
  "repositoryRootPath": "/Users/adrien/projects/my-org/adrien-python-test"
}
